[TextureMapper] Remove TextureMapperGLHeaders.h

carlosgcampos opened the pull request:
#### db6f3445bc47947748d9ded549a402a6550a8445
<pre>
[TextureMapper] Remove TextureMapperGLHeaders.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=265442">https://bugs.webkit.org/show_bug.cgi?id=265442</a>

Reviewed by NOBODY (OOPS!).

Just include the GL headers from cpp files only, since it can conflict
between epoxy, angle and system headers.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.cpp:
(WebCore::GstVideoFrameHolder::platformLayerBuffer):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::pushTextureToCompositor):
(WebCore::MediaPlayerPrivateGStreamer::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/ClipStack.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperGCGLPlatformLayer.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperGLHeaders.h: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.cpp:
(WebCore::TextureMapperPlatformLayerBuffer::TextureMapperPlatformLayerBuffer):
(WebCore::TextureMapperPlatformLayerBuffer::canReuseWithoutReset):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::getAvailableBuffer):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.h:
(WebCore::TextureMapperShaderProgram::programID const):
* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/db6f3445bc47947748d9ded549a402a6550a8445

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7374 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30102 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31350 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29301 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9499 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4734 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26283 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24696 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5303 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/notification (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5454 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25700 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32688 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26299 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26143 "Found 1 new test failure: fullscreen/full-screen-layer-dump.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31698 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3591 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29478 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7038 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5888 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5939 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii reviewed (commented):
If you want to hide OpenGL header, your merging TextureMapperGL is a mistake. It's a kind of pimpl idiom.
https://bugs.webkit.org/show_bug.cgi?id=263961

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapper.h:
```diff
@@ -65,11 +64,11 @@ class TextureMapper {
+    void drawTexturePlanarYUV(const std::array<unsigned, 3>& textures, const std::array<float, 16>& yuvToRgbMatrix, OptionSet<TextureMapperFlags>, const FloatRect& targetRect, const TransformationMatrix& modelViewMatrix, float opacity, std::optional<unsigned> alphaPlane, AllEdgesExposed = AllEdgesExposed::Yes);
+    void drawTextureSemiPlanarYUV(const std::array<unsigned, 2>& textures, bool uvReversed, const std::array<float, 16>& yuvToRgbMatrix, OptionSet<TextureMapperFlags>, const FloatRect& targetRect, const TransformationMatrix& modelViewMatrix, float opacity, AllEdgesExposed = AllEdgesExposed::Yes);
+    void drawTexturePackedYUV(unsigned texture, const std::array<float, 16>& yuvToRgbMatrix, OptionSet<TextureMapperFlags>, const FloatRect& targetRect, const TransformationMatrix& modelViewMatrix, float opacity, AllEdgesExposed = AllEdgesExposed::Yes);
+    void drawTextureExternalOES(unsigned texture, OptionSet<TextureMapperFlags>, const FloatRect&, const TransformationMatrix& modelViewMatrix, float opacity);
```
This change doesn't look acceptable.
You need pimpl idiom or restoring TextureMapper abstact class.

carlosgcampos replied:
Why? I don't see the problem. Note that this is working for you because you use angle for everything.

carlosgcampos replied:
Let's use `GraphicsTypesGL.h` like the rest of WebCore does, then.

fujii replied:
GraphicsTypesGL.h is for GraphicsContextGL.
TextureMaper is using system OpenGL while GraphicsContextGL (WebGL) is using ANGLE in GTK and WPE ports.
The previous is better design before 270048@main merged TextureMaperGL in order to hide OpenGL headers.
https://bugs.webkit.org/show_bug.cgi?id=263961
We should use pimpl idiom or TextureMaper abstruct class to hide OpenGL headers.

carlosgcampos replied:
> GraphicsTypesGL.h is for GraphicsContextGL. 

Well, it's used in other places, in some cases we even include GraphicsContextGL.h when we really just need GraphicsTypesGL.h. But the idea is the same in any case, it uses its own types matching the GL ones to avoid including the GL headers.

> TextureMaper is using system OpenGL while GraphicsContextGL (WebGL) is using ANGLE in GTK and WPE ports.

Yes, and that's still the case, the goal of this patch is not to change that, only to avoid including the GL headers from the header classes.

> The previous is better design before 270048@main merged TextureMaperGL in order to hide OpenGL headers.

That was not the intention of the previous design, and it didn'0t work either, because TextureMapperGL.h already indluded the GL header and was included directly in some other files. The previous design comes from the times when it was possible to use TextureMapper with a software based path. That was removed and we kept the abstraction that was no longer needed.

>  https://bugs.webkit.org/show_bug.cgi?id=263961 We should use pimpl idiom or TextureMaper abstruct class to hide OpenGL headers.

I don't understand why, and I don't see how adding back the abstract class helps.

fujii replied:
> because TextureMapperGL.h already indluded the GL header and was included directly in some other files.

A callsite doesn't need to include TextureMapperGL.h. It should use `TextureMapper::create` factory class method instead.

> I don't understand why, and I don't see how adding back the abstract class helps.

abstract class can be considered as pimpl idiom.
I'm suggesting pimp idiom if you don't like the idea of abstract class.

carlosgcampos replied:
What I still don't know is why. An abstract class or a pimpl tries to hide the implementation details, but this is not the goal of this patch, this is a GL class and it's ok, we just want to avoid including the GL headers from the header, because of the way epoxy, angle and system headers work.

fujii replied:
This change requires us to use `GCGLenum` and `GCGLuint` types in TextureMapper.h. The prefix `GCGL` is a addriviation of `GraphicsContextGL`. It's a class for WebGL. It's not relavant to TextureMapper.

carlosgcampos replied:
That was probably the initial intention, but now it's used in other parts not related to WebGL. Anyway, let's go back to the previous version and use int, unsigned, float. I still don't understand the problem of using those types when it's documented that GLint is int, GLuint unsigned int, etc.

fujii replied:
Other parts? Which code?
Your design looks pecuriler. The previous TextureMapperGL is a better design.
WebKit project is fastidious to even a single whitespace.

carlosgcampos replied:
> Other parts? Which code? 

Ok, I grepped and found several places where `GraphicsTypesGL.h` is used, but it seems we only use the GC types in GLContext, which can easily changed too. I don't want to use the GC types either, I was just trying to understand what you wanted.

> Your design looks pecuriler. The previous TextureMapperGL is a better design.

Again, the previous design had a different goal, it was possible to have a software based texture mapper. I don't want to hide the GL implementation details, we want TextureMapper to be GL, we just need to avoid GL headers inclusion from header files, because epoxy, angle and system headers can't be included together. So the only thing we need here is using unsigned instead of GLuint, which is exactly the same thing and documented as such. You might get the same problem if you didn't use ANGLE for everything.

> WebKit project is fastidious to even a single whitespace.

I'm sorry, but I don't understand what you want. You keep saying that previous design was better, but you haven't answered why yet, or what the problem of using unsigned instead of GLuint is.

fujii replied:
No one says there is a problem in both designs. Your design looks pecuriler. We should move to webkit-dev.

carlosgcampos committed:
[TextureMapper] Remove TextureMapperGLHeaders.h
https://bugs.webkit.org/show_bug.cgi?id=265442

Reviewed by NOBODY (OOPS!).

Just include the GL headers from cpp files only, since it can conflict
between epoxy, angle and system headers.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameHolder.cpp:
(WebCore::GstVideoFrameHolder::platformLayerBuffer):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::pushTextureToCompositor):
(WebCore::MediaPlayerPrivateGStreamer::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/holepunch/MediaPlayerPrivateHolePunch.cpp:
(WebCore::MediaPlayerPrivateHolePunch::pushNextHolePunchBuffer):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.cpp:
(WebCore::BitmapTexture::BitmapTexture):
* Source/WebCore/platform/graphics/texmap/BitmapTexture.h:
* Source/WebCore/platform/graphics/texmap/ClipStack.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperGCGLPlatformLayer.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperGLHeaders.h: Removed.
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.cpp:
(WebCore::TextureMapperPlatformLayerBuffer::TextureMapperPlatformLayerBuffer):
(WebCore::TextureMapperPlatformLayerBuffer::canReuseWithoutReset):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerBuffer.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyDMABuf.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.cpp:
(WebCore::TextureMapperPlatformLayerProxyGL::getAvailableBuffer):
* Source/WebCore/platform/graphics/texmap/TextureMapperPlatformLayerProxyGL.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.cpp:
* Source/WebCore/platform/graphics/texmap/TextureMapperShaderProgram.h:
(WebCore::TextureMapperShaderProgram::programID const):
* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
